Update UI to handle Fetching Product Status with Action Scheduler Jobs

jorgemd24 opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Part of https://github.com/woocommerce/google-listings-and-ads/issues/2146 & https://github.com/woocommerce/google-listings-and-ads/issues/2250

While we're refactoring the backend logic for the product stats and moving it to the Action Scheduler Job, we also need to update the UI to reflect these changes. This involves adjusting how the UI behaves when the job is scheduled or running and the product stats aren't available yet. Hence, we'll need to make a call to `wc/gla/mc/product-statistics/` every 30 seconds to check if the stats calculation has been completed.


https://github.com/woocommerce/google-listings-and-ads/assets/2488994/e4aff324-1074-4513-b645-f80b0deb18e9



In summary, this PR introduces the following:

- It will continuously call `wc/gla/mc/product-statistics/` every 30 seconds until the `update_merchant_product_statuses` job is finished and the stats can be retrieved.

Example response when loading:
```
{
    "timestamp": 1710711097,
    "statistics": null,
    "scheduled_sync": 0,
    "loading": true,
    "error": null
}
```
Example response when it is completed:

```
{
    "timestamp": 1710711097,
    "statistics": {
      "active" : 10
      "disapproved" : 1
      "expiring" : 2 
      "not_synced" : 3
      "pending" : 4
    },
    "scheduled_sync": 0,
    "loading": false,
    "error": null
}
```

2. Once the fetching process is completed, it will invalidate the Product Feed Query, ensuring that the Product MC Statuses are updated.

### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Make a call to `GET wp-json/wc/gla/mc/product-statistics/refresh`.
2. Navigate to `wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fproduct-feed` or refresh the product feed page.
3. Monitor the network requests and observe that the network request `wc/gla/mc/product-statistics` is fetched every 30 seconds until the loading status becomes false.
4. Verify that the product feed request is issued.
5. Confirm that the Overview Stats are loaded.
6. Ensure that the product feed displays the updated MC statuses.


### Additional details:

I was doubting whether it would be better to use an AppSpinner for each stats value or simply use the `SummaryListPlaceholder`. In the end, I decided to go with the placeholder because it maintains a more consistent look with what we currently have. However, I'm open to hearing other opinions.

AppSpinner:

https://github.com/woocommerce/google-listings-and-ads/assets/2488994/5c5b50c1-0e80-49e9-9771-1f000aa98621

PlaceHolder:

![image](https://github.com/woocommerce/google-listings-and-ads/assets/2488994/36fd81a6-d8e5-4b2f-aa53-e47d57600816)


<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

>

jorgemd24 committed:
Refactor useMCProductStatistics

jorgemd24 committed:
Remove useProductFeedPrefetch

jorgemd24 committed:
Check if hasFinishedResolution in SyncStatus

jorgemd24 committed:
Add a app-spinner for SummaryNumber Product Stats

jorgemd24 committed:
Fix undefined index error

jorgemd24 committed:
Adjust js tests

jorgemd24 committed:
Fix issue if invalidating the first call

jorgemd24 committed:
Refactor useEffect

jorgemd24 committed:
Stop the countdown

jorgemd24 committed:
Stop counter if callCount > 0

jorgemd24 committed:
Use SummaryListPlaceholder instead of AppSpinner

jorgemd24 committed:
Add tests for useMCProductStatistics

jorgemd24 committed:
Add describe

jorgemd24 committed:
Add tests for Product Statistics

jorgemd24 committed:
Show sync in progress when populating the stats

jorgemd24 committed:
Update ProductStatistics types

jorgemd24 committed:
Invalidate resolution for getMCProductFeed

jorgemd24 committed:
Add test for invalidateResolutionForStoreSelector

jorgemd24 committed:
Update JSDoc

jorgemd24 committed:
Remove unused caption

jorgemd24 committed:
Cover case where there is no data

jorgemd24 committed:
Add missing caption

ianlin commented on js/src/hooks/useMCProductStatistics.js:
```diff
@@ -1,15 +1,60 @@
+		useAppSelectDispatch( 'getMCProductStatistics' );
+
+	const isLoading = hasFinishedResolution && data?.loading ? true : false;
+	const hasStats = hasFinishedResolution && ! data?.loading ? true : false;
```
Maybe we could just use `isLoading` and remove `hasStats`? Or we could check if `data.statistics` had actual data.

jorgemd24 replied:
Thanks for the suggestion, I changed a now we are checking if `data.statistics` has data. Also, I changed the response from the API, so instead of an empty array, it returns null for the statistics. See here: https://github.com/woocommerce/google-listings-and-ads/pull/2318/commits/03ceb657f447220226c33bda30999b28a779dc89

ianlin commented on js/src/hooks/useMCProductStatistics.js:
```diff
@@ -1,15 +1,60 @@
+
+		if ( isLoading && second === 0 && callCount > 0 ) {
+			invalidateResolution();
+		}
```
We could combine this two conditions like:

```suggestion
		if ( isLoading && second === 0 ) {
			 startCountdown( 30 );
			 if ( callCount > 0 ) {
			   invalidateResolution();
			 }
		}
```

ianlin replied:
Also probably might be helpful to add some comments, e.g. `Start counter when...` and `Clear the product statistic data...`

jorgemd24 replied:
I made an update and also renamed the constant from `isLoading` to `isCalculatingStats`. I believe this change clarifies that the loading belongs to the calculation process rather than indicating that the HTTP request is in a "loading" state.

ianlin commented on js/src/product-feed/product-statistics/status-box/sync-status.js:
```diff
@@ -66,7 +67,14 @@ function getSyncResult( {
+		'getMCProductStatistics'
+	);
+
+	if ( ! hasFinishedResolution || ( hasFinishedResolution && ! data ) ) {
```
It could be simplified to:

```suggestion
	if ( ! hasFinishedResolution || ! data ) {
```

ianlin commented on js/src/product-feed/product-statistics/index.js:
```diff
@@ -44,10 +53,8 @@ const ProductStatistics = () => {
-				) }
-				{ hasFinishedResolution && data && (
+				{ isLoading && <SummaryListPlaceholder numberOfItems={ 5 } /> }
+				{ hasFinishedResolution && ! data?.loading && (
```
We could probably define `hasStats` variable like above comment and check if there is actual data.

ianlin reviewed (commented):
Thanks for the change. It worked perfectly on my test environment, I only just needed to manually delay the job a little bit otherwise the request was too fast to complete. I left some comments that we could have some discussions on.

ianlin commented:
> I decided to go with the placeholder because it maintains a more consistent look with what we currently have. However, I'm open to hearing other opinions.

I think using AppSpinner would be a bit better IMO. As waiting for 30 seconds with the static summary placeholder would made me think something went wrong. But spinners spinning for 30 seconds might have the same feeling 😅.

Maybe either have something saying it is fetching the data, or reduce the interval to maybe 10 seconds might help?

jorgemd24 committed:
If no stats sent null instead of an empty array

jorgemd24 committed:
Simplify logic conditions and check for data.statistics

jorgemd24 committed:
Simplify logic condition for SyncStatus

jorgemd24 committed:
Set startCountdown every 15s

jorgemd24 committed:
Add AppSpinners when loading stats

jorgemd24 committed:
Adjust tests for the AppSpinner when loading stats

jorgemd24 committed:
Js linting .scss

jorgemd24 commented:
> I think using AppSpinner would be a bit better IMO. As waiting for 30 seconds with the static summary placeholder would made me think something went wrong. But spinners spinning for 30 seconds might have the same feeling

Yes, I have the same feeling, I switched it to AppSpinner and shortened the time to 15 seconds. I don't want to reduce the time too much because for a merchant with thousands of products, the jobs could take longer to complete all the calculations.

jorgemd24 commented:
@ianlin thanks for your time reviewing this PR.

 I addressed all your suggestions and replied to your comments! Could you please take another look? Thanks!

jorgemd24 committed:
Only load syncStatus if statistics is loaded

jorgemd24 committed:
Simplify SummaryNumber value

ianlin commented on js/src/hooks/useMCProductStatistics.js:
```diff
@@ -1,15 +1,64 @@
+			// If the job is still processing the data, invalidate the resolution/refetch the data to get the latest status.
+			if ( callCount > 0 ) {
+				invalidateResolution();
+			}
```
I feel like the comment would be more accurate if added something like `...the countdown has started...`.

ianlin reviewed (approved):
Thanks for the extra changes @jorgemd24. Using AppSpinner and reducing the countdown time is better regarding UX. I tested again and everything worked well, just left a non-blocker comment. LGTM.

jorgemd24 committed:
Adjust jsdoc for startCountdown
